src/auxiliary/math.jl

```diff
@@ -207,4 +207,70 @@ Return `x` if `x` is negative, else zero. In other words, return
 @inline function negative_part(x)
     return min(x, zero(x))
 end
+
+"""
+    stolarsky_mean(x, y)
+
+Compute an instance of a weighted Stolarsky mean of the form
+
+    stolarsky_mean(x, y, gamma) = (gamma - 1)/gamma * (y^gamma - x^gamma) / (y^(gamma-1) - x^(gamma-1))
+
+where `gamma > 1`.
+
+Problem: The formula above has a removable singularity at `x == y`. Thus,
+some care must be taken to implement it correctly without problems or loss
+of accuracy when `x ≈ y`. Here, we use the approach proposed by
+Winters et al. (2020).
+Set f = (y - x) / (y + x) and g = gamma (for compact notation).
+Then, we use the expansions
+
+    ((1+f)^g - (1-f)^g) / g = 2*f + (g-1)(g-2)/3 * f^3 + (g-1)(g-2)(g-3)(g-4)/60 * f^5 + O(f^7)
+
+and
+
+    ((1+f)^(g-1) - (1-f)^(g-1)) / (g-1) = 2*f + (g-2)(g-3)/3 * f^3 + (g-2)(g-3)(g-4)(g-5)/60 * f^5 + O(f^7)
+
+Inserting the first few terms of these expansions and performing polynomial long division
+we find that
+
+    stolarsky_mean(x, y, gamma) ≈ (y + x) / 2 * (1 + (g-2)/3 * f^2 - (g+1)(g-2)(g-3)/45 * f^4 + (g+1)(g-2)(g-3)(2g(g-2)-9)/945 * f^6)
+
+Since divisions are usually more expensive on modern hardware than
+multiplications (Agner Fog), we try to avoid computing two divisions. Thus,
+we use
+
+    f^2 = (y - x)^2 / (x + y)^2
+        = (x * (x - 2 * y) + y * y) / (x * (x + 2 * y) + y * y)
+
+Given ε = 1.0e-4, we use the following algorithm.
+
+    if f^2 < ε
+      # use the expansion above
+    else
+      # use the direct formula (gamma - 1)/gamma * (y^gamma - x^gamma) / (y^(gamma-1) - x^(gamma-1))
+    end
+
+# References
+- Andrew R. Winters, Christof Czernik, Moritz B. Schily & Gregor J. Gassner (2020)
+  Entropy stable numerical approximations for the isothermal and polytropic
+  Euler equations
+  [DOI: 10.1007/s10543-019-00789-w](https://doi.org/10.1007/s10543-019-00789-w)
+- Agner Fog.
+  Lists of instruction latencies, throughputs and micro-operation breakdowns
+  for Intel, AMD, and VIA CPUs.
+  [https://www.agner.org/optimize/instruction_tables.pdf](https://www.agner.org/optimize/instruction_tables.pdf)
+"""
+@inline function stolarsky_mean(x, y, gamma)
+    epsilon_f2 = 1.0e-4
+    f2 = (x * (x - 2 * y) + y * y) / (x * (x + 2 * y) + y * y) # f2 = f^2
+    if f2 < epsilon_f2
+        # convenience coefficients
+        c1 = (gamma - 2) / 3
+        c2 = -(gamma + 1) * (gamma - 2)* (gamma - 3) / 45
+        c3 = (gamma + 1) * (gamma - 2) * (gamma - 3)* (2 * gamma * (gamma - 2) - 9) / 945
```

Comment:
```suggestion
        c1 = (1 / 3) * (gamma - 2)
        c2 = -(1 / 45) * (gamma + 1) * (gamma - 2)* (gamma - 3)
        c3 = (1 / 945) * (gamma + 1) * (gamma - 2) * (gamma - 3)* (2 * gamma * (gamma - 2) - 9)
```
This should avoid some relatively expensive divisions

Reply:
Because the division is done at compile time and then stored?

Reply:
If it further helps, we could make this (potentially) computationally cheaper by reusing the already computing coefficients. This would make the formula unlike how it is written on paper but could save a few multiplications overall. Here is what I mean:
```julia
c1 = (1 / 3) * (gamma - 2)
c2 = -(1 / 15) * (gamma + 1) * (gamma - 3) * c1
c3 = -(1 / 21) * (2 * gamma *(gamma - 2) - 9) * c2
```

Reply:
> Because the division is done at compile time and then stored?

Yes:

```julia
julia> foo(x) = x / 3
foo (generic function with 1 method)

julia> @code_native debuginfo=:none syntax=:intel foo(1.0)
        .text
        .file   "foo"
        .section        .rodata.cst8,"aM",@progbits,8
        .p2align        3                               # -- Begin function julia_foo_98
.LCPI0_0:
        .quad   0x4008000000000000              # double 3
        .text
        .globl  julia_foo_98
        .p2align        4, 0x90
        .type   julia_foo_98,@function
julia_foo_98:                           # @julia_foo_98
        .cfi_startproc
# %bb.0:                                # %top
        push    rbp
        .cfi_def_cfa_offset 16
        .cfi_offset rbp, -16
        mov     rbp, rsp
        .cfi_def_cfa_register rbp
        movabs  rax, offset .LCPI0_0
        vdivsd  xmm0, xmm0, qword ptr [rax]
        pop     rbp
        .cfi_def_cfa rsp, 8
        ret
.Lfunc_end0:
        .size   julia_foo_98, .Lfunc_end0-julia_foo_98
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits

julia> bar(x) = (1 / 3) * x
bar (generic function with 1 method)

julia> @code_native debuginfo=:none syntax=:intel bar(1.0)
        .text
        .file   "bar"
        .section        .rodata.cst8,"aM",@progbits,8
        .p2align        3                               # -- Begin function julia_bar_123
.LCPI0_0:
        .quad   0x3fd5555555555555              # double 0.33333333333333331
        .text
        .globl  julia_bar_123
        .p2align        4, 0x90
        .type   julia_bar_123,@function
julia_bar_123:                          # @julia_bar_123
        .cfi_startproc
# %bb.0:                                # %top
        push    rbp
        .cfi_def_cfa_offset 16
        .cfi_offset rbp, -16
        mov     rbp, rsp
        .cfi_def_cfa_register rbp
        movabs  rax, offset .LCPI0_0
        vmulsd  xmm0, xmm0, qword ptr [rax]
        pop     rbp
        .cfi_def_cfa rsp, 8
        ret
.Lfunc_end0:
        .size   julia_bar_123, .Lfunc_end0-julia_bar_123
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
```

Reply:
Implemented efficient version of @andrewwinters5000 .